File: packages/ios/Sources/Components/FeedbackCard/FeedbackCard.swift
import SwiftUI

public struct KozmosFeedbackCard: View {
    @Environment(\.kozmosAnalytics) private var trackEvent

    private let title: String
    private let description: String
    private let isSubmitting: Bool
    private let successMessage: String
    private let surface: KozmosSurfaceStyle
    private let onSubmitFeedback: ((Int, String) -> Void)?

    @State private var rating: Int
    @State private var comment: String
    @State private var submitted: Bool

    /// `surface` is what the card is made of, as React's `surface` prop: `.solid`
    /// (the default) or `.glass`, for a card over the map. Until 2026-09-22
    /// SwiftUI drew it solid only.
    public init(
        title: String = "Rate your experience",
        description: String = "How was your navigation today?",
        isSubmitting: Bool = false,
        successMessage: String = "Thank you for the feedback!",
        initialRating: Int = 0,
        initialComment: String = "",
        surface: KozmosSurfaceStyle = .solid,
        onSubmitFeedback: ((Int, String) -> Void)? = nil
    ) {
        self.surface = surface
        self.title = title
        self.description = description
        self.isSubmitting = isSubmitting
        self.successMessage = successMessage
        self.onSubmitFeedback = onSubmitFeedback
        self._rating = State(initialValue: initialRating)
        self._comment = State(initialValue: initialComment)
        self._submitted = State(initialValue: false)
    }

    public var body: some View {
        VStack(spacing: KozmosDimensions.primitivesLayoutSpacing200) {
            if submitted {
                VStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                    Image(systemName: "checkmark.circle.fill")
                        .font(.system(size: 44, weight: .semibold))
                        .foregroundColor(KozmosColors.primitivesColorsEmotionalSuccess600)

                    Text(successMessage)
                        .font(.headline.weight(.semibold))
                        .foregroundColor(KozmosColors.primitivesColorsForeground100)
                        .multilineTextAlignment(.center)
                }
                .frame(maxWidth: .infinity, minHeight: 148)
            } else {
                VStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
                    Text(title)
                        .font(.headline.weight(.semibold))
                        .foregroundColor(KozmosColors.primitivesColorsForeground100)

                    Text(description)
                        .font(KozmosTypography.subheadline)
                        .foregroundColor(KozmosColors.primitivesColorsForeground500)
                        .multilineTextAlignment(.center)
                }

                KozmosRating(value: $rating)

                KozmosTextarea(text: $comment, placeholder: "Tell us more about your experience...")
                    .frame(minHeight: 96)

                KozmosButton(
                    isSubmitting ? "Submitting..." : "Submit Feedback",
                    isDisabled: rating == 0 || isSubmitting
                ) {
                    trackEvent(KozmosAnalyticsEvent(eventName: "feedback_submitted", component: "FeedbackCard", properties: ["rating": String(rating)]))
                    onSubmitFeedback?(rating, comment)
                    submitted = true
                }
                .frame(maxWidth: .infinity)
            }
        }
        .padding(KozmosDimensions.primitivesLayoutSpacing300)
        .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous))
        // The solid surface React's card sits on by default: the background
        // with the subtle border. It was the background at 90 % under a
        // near-black hairline at 8 % until 2026-09-22.
        .kozmosSurface(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous), style: surface)
        .kozmosElevation(KozmosShadows.semanticsElevationOverlay)
    }
}

File: packages/android/src/main/java/com/kozmos/components/FeedbackCard/FeedbackCard.kt
package com.kozmos.components.feedbackcard

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.kozmos.components.button.KozmosButton
import com.kozmos.components.rating.KozmosRating
import com.kozmos.components.textarea.KozmosTextarea
import com.kozmos.providers.KozmosAnalyticsEvent
import com.kozmos.providers.LocalKozmosAnalytics
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.components.surface.KozmosSurfaceDefaults
import com.kozmos.components.surface.KozmosSurfaceStyle

@Composable
fun KozmosFeedbackCard(
    modifier: Modifier = Modifier,
    title: String = "Rate your experience",
    description: String = "How was your navigation today?",
    isSubmitting: Boolean = false,
    successMessage: String = "Thank you for the feedback!",
    onSubmitFeedback: ((Int, String) -> Unit)? = null
) {
    val trackEvent = LocalKozmosAnalytics.current
    var rating by remember { mutableStateOf(0) }
    var comment by remember { mutableStateOf("") }
    var submitted by remember { mutableStateOf(false) }

    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel),
        // The solid surface React's card sits on by default, themed: the
        // background with the subtle border. It was the background at 90 %
        // under a near-black hairline at 8 % until 2026-09-22.
        color = KozmosSurfaceDefaults.tint(KozmosSurfaceStyle.Solid),
        tonalElevation = 6.dp,
        shadowElevation = 12.dp,
        border = KozmosSurfaceDefaults.border(KozmosSurfaceStyle.Solid)
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing200),
            modifier = Modifier
                .fillMaxWidth()
                .padding(KozmosDimensions.primitivesLayoutSpacing300)
        ) {
            if (submitted) {
                Icon(
                    imageVector = Icons.Default.CheckCircle,
                    contentDescription = null,
                    tint = KozmosColors.primitivesColorsEmotionalSuccess600,
                    modifier = Modifier.size(48.dp)
                )
                Text(
                    text = successMessage,
                    style = MaterialTheme.typography.titleMedium,
                    color = KozmosColors.primitivesColorsForeground100,
                    textAlign = TextAlign.Center
                )
            } else {
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleMedium,
                    color = KozmosColors.primitivesColorsForeground100,
                    textAlign = TextAlign.Center
                )
                Text(
                    text = description,
                    style = MaterialTheme.typography.bodyMedium,
                    color = KozmosColors.primitivesColorsForeground500,
                    textAlign = TextAlign.Center
                )

                KozmosRating(value = rating, onValueChange = { rating = it })

                KozmosTextarea(
                    value = comment,
                    onValueChange = { comment = it },
                    placeholder = "Tell us more about your experience...",
                    modifier = Modifier.heightIn(min = 96.dp)
                )

                KozmosButton(
                    onClick = {
                        trackEvent(KozmosAnalyticsEvent(component = "FeedbackCard", eventName = "feedback_submitted", properties = mapOf("rating" to rating.toString())))
                        onSubmitFeedback?.invoke(rating, comment)
                        submitted = true
                    },
                    enabled = rating > 0 && !isSubmitting,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Text(if (isSubmitting) "Submitting..." else "Submit Feedback")
                }
            }
        }
    }
}
